nostt.1
.Dd August 5, 2023
.Dt NOSTT 1
.Os
.Sh NAME
.Nm nostt
.Nd command line NOS Teletekst reader
.Sh SYNOPSIS
.Nm
.Op Ar page
.Sh DESCRIPTION
Displays the requested NOS Teletekst page on the command line.
Pages are identified by a three-digit number optionally followed by a subpage,
for example 100 or 302-2.
All subpages are printed if not specified.
.Pp
If no
.Ar page
argument is given,
.Nm
runs interactively.
.Sh ENVIRONMENT
.Bl -tag -width LC_MONETARY
.It Ev NO_COLOR
When set to 1, color output is disabled.
.It Ev CLICOLOR
When set to 0, color output is disabled.
.It Ev CLICOLOR_FORCE
When set to 1, color output is enabled
even output is redirect to a pipe or file.
Note that
.Ev NO_COLOR=1 and
.Ev CLICOLOR=0
take precedence.
.El
.Sh SEE ALSO
Project page:
.Lk https://github.com/sjmulder/nostt
.Pp
Official NOS Teletekst web client:
.Lk https://nos.nl/teletekst
.Sh AUTHORS
.An Sijmen J. Mulder
.Aq Mt ik@sjmulder.nl .
.Sh BUGS
Teletext's 6-cell (2x3) block characters are displayed as % characters
as they cannot be accurately reproduced in a terminal without special fonts.
(The NOS viewer and API uses a custom font with these characters in the
.Qq private use
0xF000 Unicode range.)
.Pp
Color ouptut is disabled by default in MinTTY on Windows.
This is because MinTTY redirects standard output.
Set
.Ev CLICOLOR_FORCE
to 1 to force color output.


CHANGES.md
1.3 (2023-08-03)
----------------
 - Fixed: typo and style in man page (thanks Omar Polo).
 - New: Dutch translation of the man page.
 - New: unveil() support on OpenBSD (thanks Tim van der Molen).
 - New: support for NO_COLOR=1 (http://no-color.org).
 - New: non-canonical CMake build file.
 - Change: migrated builds from sr.ht to GitHub Actions.

1.2.2 (2022-06-15)
------------------
 - Fixed: API error caused by use of old HTTP endpoint. Switched to
   same endpoint but HTTPS.

1.2.1 (2020-07-25)
------------------
 - New: Dutch translation of README.md.
 - Fixed: interactive prompt not printed with musl libc.
 - Fixed: typos in README.md.
 - Change: migrated builds from Travis to sr.ht.

1.2 (2020-01-02)
----------------
 - Fixed: build warning from mismatched callback signature.
 - Fixed: only first column printed with colors off.
 - New: type 'q' at the prompt to quit.
 - New: next page suggested after last subpage (e.g. [101] after [101-4]).
 - Change: drop /usr/local from default include and lib search path. System
   search paths aren't for us to decide.

1.1.1 (2018-12-06)
------------------
 - Fixed: wrong implicit permissions in `make install`.

1.1 (2018-11-12)
----------------
 - New: 'make check' target.
 - Removed: -i option (interactive mode). Interactive mode is now used if no
   page argument is given.
 - Removed: -G option (color ouptut). Color output is the default now for
   interactive terminals, but can be disabled with environment CLICOLOR=0 or
   forced with CLICOLOR_FORCE=1.
 - Changed: default CFLAGS and LDFLAGS, in particular adding /usr/local
   search paths.
 - Changed: make [un]install now respects DESTDIR. prefix, bindir, and
   man1dir have been replaced with PREFIX and MANPREFIX.
 - Fixed: several portability issues.

1.0 (2018-03-04)
----------------
Initial release.


nostt.c
/* nostt.c - Copyright (c) 2018, Sijmen J. Mulder (see LICENSE.md) */

#define USAGE	"usage: nostt [page]"

#define _WITH_GETLINE

#include <unistd.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <wchar.h>
#include <locale.h>
#include "api.h"
#include "compat.h"

#ifdef _WIN32
# include <windows.h>
#endif

#ifndef COMPAT_ERR
# include <err.h>
#endif

static void
enablecolor()
{
#ifdef _WIN32
	HANDLE	hout;
	DWORD	mode	= 0;

	hout = GetStdHandle(STD_OUTPUT_HANDLE);
	GetConsoleMode(hout, &mode);
	/* ENABLE_VIRTUAL_TERMINAL_PROCESSING */
	SetConsoleMode(hout, mode|4);
#endif
}

static int
enveq(const char *name, const char *val)
{
	char	*actual;

	return (actual = getenv(name)) && strcmp(actual, val) == 0;
}

static void
putcell_color(struct ttpage *page, int line, int col)
{
	struct ttattrs	*attrs;
	struct ttattrs	*prevattrs;
	int		 attrsflag;

	attrs = &page->attrs[line][col];
	if (!col)
		attrsflag = 1;
	else {
		prevattrs = &page->attrs[line][col-1];
		attrsflag = memcmp(attrs, prevattrs, sizeof(*attrs));
	}

	if (attrsflag)
		printf("\033[%d;%dm", 30 + attrs->fg, 40 + attrs->bg);

	putwchar(page->chars[line][col]);
}

static const char *
prompt(const char *suggestion)
{
	static char	*input = NULL;
	static ssize_t	 len;
	static size_t	 cap;

	while (1) {
		if (suggestion)
			printf("page [%s]? ", suggestion);
		else
			printf("page? ");
		fflush(stdout);

		if ((len = getline(&input, &cap, stdin)) == -1)
			return NULL;
		if (len && input[len-1] == '\n')
			input[--len] = '\0';

		if (!strcmp("q", input))
			return NULL;
		else if (len)
			return input;
		else
			return suggestion;
	}
}

int
main(int argc, char **argv)
{
	const char	*id;
	struct ttpage 	 page;
	enum tterr	 ret;
	int		 withcolor	= 0;
	int		 interactive	= 0;
	int		 line, col;

#ifdef __OpenBSD__
	if (unveil("/etc/ssl/cert.pem", "r") == -1)
		err(1, "unveil: /etc/ssl/cert.pem");
	if (unveil("/etc/ssl/openssl.cnf", "r") == -1)
		err(1, "unveil: /etc/ssl/openssl.cnf");
	if (unveil("/usr/share/locale", "r") == -1)
		err(1, "unveil: /usr/share/locale");
	if (pledge("stdio rpath inet dns", NULL) == -1)
		err(1, "pledge");
#endif

	(void) argc;

	argv0 = *argv;
	setlocale(LC_ALL, "");

	if (argv[1]) {
		if (argv[2] || *argv[1] == '-')
			errx(1, USAGE);
		id = argv[1];
	} else {
		id = "100";
		interactive = 1;
	}

	withcolor =
	    !enveq("NO_COLOR", "1") &&
	    !enveq("CLICOLOR", "0") &&
	    (enveq("CLICOLOR_FORCE", "1") || isatty(STDOUT_FILENO));

	if (withcolor)
		enablecolor();

	while (1) {
		ret = tt_get(id, &page);
		if (ret != TT_OK) {
			if (!interactive)
				errx(1, "%s", tt_errstr(ret));

			warnx("%s", tt_errstr(ret));
			puts("(q or ^C to exit)\n");
			if (!(id = prompt(NULL)))
				return 0;
			continue;
		}

		if (withcolor) {
			for (line = 0; line < TT_NLINES; line++) {
				for (col = 0; col < TT_NCOLS; col++)
					putcell_color(&page, line, col);
				puts("\033[0m");
			}
		} else {
			for (line = 0; line < TT_NLINES; line++)
				printf("%.*ls\n", TT_NCOLS,
				    page.chars[line]);
		}

		if (interactive) {
			putchar('\n');
			id = prompt(
			    *page.nextsub ? page.nextsub :
			    *page.nextpage ? page.nextpage : NULL);
			if (!id)
				return 0;
		} else if (id == argv[0] && strchr(id, '-')) {
			/* Only print requested subpage */
			return 0;
		} else if (*page.nextsub) {
			putchar('\n');
			id = page.nextsub;
		} else
			return 0;
	}
}
